File: src/Butil/Bit.Butil/Publics/StorageManager.cs
using System.Diagnostics.CodeAnalysis;
using System.Threading.Tasks;
using Microsoft.JSInterop;

namespace Bit.Butil;

/// <summary>
/// Wraps <see href="https://developer.mozilla.org/en-US/docs/Web/API/StorageManager">navigator.storage</see>.
/// </summary>
public class StorageManager(IJSRuntime js)
{
    /// <summary>True when the runtime exposes <c>navigator.storage</c>.</summary>
    /// <remarks>
    /// During prerender/SSR (no JS runtime) this returns <c>default</c> (e.g. <c>false</c>/<c>0</c>)
    /// rather than throwing, so the result can't be distinguished from a genuine value. If you
    /// branch on it, defer the read to <c>OnAfterRenderAsync</c>.
    /// </remarks>
    public ValueTask<bool> IsSupported() => js.Invoke<bool>("BitButil.storageManager.isSupported");

    /// <summary>
    /// Reports an estimate of the storage quota and current usage for the origin.
    /// </summary>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(StorageEstimate))]
    public ValueTask<StorageEstimate> Estimate() => js.Invoke<StorageEstimate>("BitButil.storageManager.estimate");

    /// <summary>True when the origin's storage is persistent (won't be evicted under pressure).</summary>
    /// <remarks>
    /// During prerender/SSR (no JS runtime) this returns <c>default</c> (e.g. <c>false</c>/<c>0</c>)
    /// rather than throwing, so the result can't be distinguished from a genuine value. If you
    /// branch on it, defer the read to <c>OnAfterRenderAsync</c>.
    /// </remarks>
    public ValueTask<bool> Persisted() => js.Invoke<bool>("BitButil.storageManager.persisted");

    /// <summary>
    /// Asks the browser to make storage persistent. The user agent decides whether to grant.
    /// </summary>
    /// <remarks>
    /// During prerender/SSR (no JS runtime) this returns <c>default</c> (e.g. <c>false</c>/<c>0</c>)
    /// rather than throwing, so the result can't be distinguished from a genuine value. If you
    /// branch on it, defer the read to <c>OnAfterRenderAsync</c>.
    /// </remarks>
    public ValueTask<bool> Persist() => js.Invoke<bool>("BitButil.storageManager.persist");
}

File: src/Butil/Bit.Butil/Scripts/storageManager.ts
var BitButil = (window as any).BitButil = (window as any).BitButil || {};

(function (butil: any) {
    butil.storageManager = {
        isSupported() { return !!(window.navigator as any).storage; },
        async estimate() {
            const sm: any = (window.navigator as any).storage;
            if (!sm?.estimate) return { quota: null, usage: null, usageDetails: [] };
            try {
                const e = await sm.estimate();
                return {
                    quota: typeof e.quota === 'number' ? e.quota : null,
                    usage: typeof e.usage === 'number' ? e.usage : null,
                    // Chromium-only per-API breakdown ("indexedDB", "caches", "serviceWorkerRegistrations",
                    // ...). Flattened to a list so .NET sees one shape rather than an open-ended object.
                    usageDetails: Object.entries(e.usageDetails || {})
                        .map(([api, bytes]) => ({ api, bytes: typeof bytes === 'number' ? bytes : 0 }))
                };
            } catch {
                return { quota: null, usage: null, usageDetails: [] };
            }
        },
        async persisted() {
            const sm: any = (window.navigator as any).storage;
            if (!sm?.persisted) return false;
            try { return await sm.persisted(); } catch { return false; }
        },
        async persist() {
            const sm: any = (window.navigator as any).storage;
            if (!sm?.persist) return false;
            try { return await sm.persist(); } catch { return false; }
        }
    };
}(BitButil));
